xds/internal/xdsclient/tests/ads_stream_flow_control_test.go

```diff
@@ -0,0 +1,623 @@
+	return &wrappedADSStream{
+		AggregatedDiscoveryService_StreamAggregatedResourcesClient: stream,
+		recvCh: make(chan struct{}, 1),
+		doneCh: doneCh,
+	}
+}
+
+func (w *wrappedADSStream) Recv() (*v3discoverypb.DiscoveryResponse, error) {
+	select {
+	case w.recvCh <- struct{}{}:
+	case <-w.doneCh:
+		return nil, errors.New("Recv() called after the test has finished")
+	}
+	return w.AggregatedDiscoveryService_StreamAggregatedResourcesClient.Recv()
+}
+
+// Overrides the function to create a new ADS stream (used by the xdsclient
+// transport), and returns a wrapped ADS stream, where the test can monitor
+// Recv() calls.
+func overrideADSStreamCreation(t *testing.T) chan *wrappedADSStream {
+	t.Helper()
+
+	adsStreamCh := make(chan *wrappedADSStream, 1)
+	origNewADSStream := xdsclientinternal.NewADSStream
+	xdsclientinternal.NewADSStream = func(ctx context.Context, cc *grpc.ClientConn) (v3adsgrpc.AggregatedDiscoveryService_StreamAggregatedResourcesClient, error) {
+		s, err := v3adsgrpc.NewAggregatedDiscoveryServiceClient(cc).StreamAggregatedResources(ctx)
+		if err != nil {
+			return nil, err
+		}
+		ws := newWrappedADSStream(s, ctx.Done())
+		select {
+		case adsStreamCh <- ws:
+		default:
+		}
+		return ws, nil
+	}
+	t.Cleanup(func() { xdsclientinternal.NewADSStream = origNewADSStream })
+	return adsStreamCh
+}
+
+// Creates an xDS client with the given bootstrap contents.
+func createXDSClient(t *testing.T, bootstrapContents []byte) xdsclient.XDSClient {
+	t.Helper()
+
+	client, close, err := xdsclient.NewForTesting(xdsclient.OptionsForTesting{
+		Name:     t.Name(),
+		Contents: bootstrapContents,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	t.Cleanup(close)
+	return client
+}
+
+// Tests ADS stream level flow control with a single resource. The test does the
+// following:
+//   - Starts a management server and configures a listener resource on it.
+//   - Creates an xDS client to the above management server, starts a couple of
+//     listener watchers for the above resource, and verifies that the update
+//     reaches these watchers.
+//   - These watchers don't invoke the onDone callback until explicitly
+//     triggered by the test. This allows the test to verify that the next
+//     Recv() call on the ADS stream does not happen until both watchers have
+//     completely processed the update, i.e invoke the onDone callback.
+//   - Resource is update on the management server, and the test verifies that
+//     the update reaches the watchers.
+func (s) TestADSFlowControl_ResourceUpdates_SingleResource(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Override the ADS stream creation.
+	adsStreamCh := overrideADSStreamCreation(t)
+
+	// Start an xDS management server.
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{})
```

Comment:
Btw, I moved some stuff to testutils in this PR (for xDS e2e tests): https://github.com/grpc/grpc-go/pull/7466/files#diff-adea373aef70697f15ed647c2268b1e8d3f87a08ad6178f7444671c67f3264b1R43